### ISSUE TITLE
Support entries & chunks from external sources

### ISSUE THREAD

**Issue by espen42** [enonic/lib-react4xp#46]
Possible today: 
- Dependency chunks can be _compiled_ from source (under `chunkDirs`, or anywhere actually), or es6-standard _imported_ in an entry, from `/node_modules/`. 
- Entries can be _compiled_ from source (`entryDirs` and `/site/`).
- Both `entryDirs` and `chunkDirs` may point anywhere relative to `/src/main/resources/react4xp/` - including folders inside `/node_modules/`. This makes it possible to use NPM for both, but only for pre-compiled sources.

Not possible today, would be nice to have:
- Using pre-compiled entries, e.g. from `/node_modules/`
- Using pre-compiled entries and chunks by URL, from a CDN etc.

**Comment by gilmarllen** [comment]
Does it mean that we can't use any external libs in our jsx files for now? I've tried to use the 'react-konva' lib in a test but a got an error:
![image](https://user-images.githubusercontent.com/18606900/79272614-12836e00-7e78-11ea-9176-a928c35f1a80.png)

```
import React, { useState } from 'react';
import { Stage, Layer, Star, Text } from 'react-konva';

const App = props => {
  return (
    <Stage width={500} height={500}>
      <Layer>
        <Rect
          x={50}
          y={50}
          width={20}
          height={20}
          fill={"red"}
        />
      </Layer>
    </Stage>
  )
}

export default props => <App {...props} />
```

**Comment by espen42** [comment]
No, that's unrelated - this issue describes a corner case that seems different from what your error appears to be. External libs should work fine _in pronciple_. However, they might use JS functionality that is not supported in the nashorn engine that runs the serverside rendering (SSR). 

The **nashorn polyfill** is a part of react4xp that aims to solve this out of the box. That should cover third party libs too. But once in a while there are cases that aren't covered. 

My guess is that your error is the SSR trying to process a JS script that contains an addEventListener call - in your own code and/or in a third party library, and that causes problems. This might be a general flaw, something that needs a fix in a future version.

In the early stages of react4xp I laid some groundwork for making it possible for developers to add their own polyfill to the nashorn engine. There has been some changes here and there in the prototype since then, and I haven't focused on it for a while - so I'm not entirely sure how easy that will be for you (it will involve ´react4xp.properties` and/or `build.gradle`, depending on your setup and version).

The SSR stage in only used to build an initial HTML string from the react components, with a set of inital props. Assuming that HTML doesn't rely on `addEventListener` (so that function is only needed on the actual activation (hydration) of your component in the browser - which should work fine), it should be possible to work around your problem by adding a do-nothing version of window.addEventListener, something like `if (typeof window.addEventListener !== 'function') { window.addEventListener = () => {}; }`. Ideally, this should be added to the nashorn polyfill. But failing that, you could try adding a line like that as early as possible in the code that's passed through the SSR.

Otherwise,  you could try to add a `clientRender:true` option as described in the tutorial, to the `render` call(s) in your controller. That way, you bypass the SSR and do the react completely old-school, working around this. Also, if you do that, you might get a more detailed/sourcemapped error message in the browser console, helping you work around SSR issues.

Good luck! :)

**Comment by gilmarllen** [comment]
I was able to use the lib adding the mentioned polyfill `if (typeof window.addEventListener !== 'function') { window.addEventListener = () => {}; }`. Thanks!

I've just posted another question on https://github.com/enonic/lib-react4xp/issues/19. I don't know if here is the best place to post these questions, so let me know if I need to move it for https://discuss.enonic.com/ or another place.

**Comment by espen42** [comment]
Just leaving a note here too: user-added nashorn polyfills appears to work fine. See example in comment: #47

**Comment by ComLock** [comment]
Will be solved by https://github.com/enonic/lib-react4xp/issues/778
